File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
private:
    void getCombination(int index, int target, vector<int> temp, vector<vector<int>> &result, vector<int>& combination) {  
        if (target == 0) {
           result.push_back(temp);
           return;
        }

        
        for(int i = index; i < combination.size();i++){  // to explore multiple option

      
            if(i > index && combination[i] == combination[i-1]) continue;
             if(combination[i] <= target) {
                     temp.push_back(combination[i]);
             getCombination(i + 1, target - combination[i], temp, result, combination);
             temp.pop_back();
             }

            
        
      }
 }

public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end()); // for consistency
        vector<int> temp;
        vector<vector<int>> result;

        getCombination(0, target, temp, result, candidates);

        return result;
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    private void findCombinationSum(int index, int[] arr, int target, List<List<Integer>> result, List<Integer> temp){
        if(target < 0) return;
        else if(target ==0){
            result.add(new ArrayList<>(temp));
        }

        for(int i=index; i<arr.length; i++){

            if(i>index && arr[i] == arr[i-1]) continue;

            if(arr[i] > target) break;
            temp.add(arr[i]);
            findCombinationSum(i+1,arr,target-arr[i],result,temp);
            temp.remove(temp.size()-1);
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        Arrays.sort(candidates);
        findCombinationSum(0,candidates,target,result,temp);
        return result;
    }
}